crates/ruff_linter/src/rules/ruff/rules/unnecessary_regular_expression.rs

```diff
@@ -173,8 +176,13 @@ impl<'a> ReFunc<'a> {
             // version
             ("sub", 3) => {
                 let repl = call.arguments.find_argument("repl", 1)?;
-                if !repl.is_string_literal_expr() {
-                    return None;
+                // make sure repl can be resolved to a string literal without
+                // backslash escapes other than ascii control characters
+                let lit = resolve_string_literal(repl, semantic)?;
+                for (c, next) in lit.value.chars().tuple_windows() {
+                    if c == '\\' && !"abfnrtv".contains(next) {
```

Comment:
Does this work for non-raw strings where `\a` is not a valid escape? Instead, it should be `\\a`.

>  if it is a string, any backslash escapes in it are processed.

Reply:
Here another case where using `replace` alters behavior:

```
>>> re.sub(r"a", r"\z", "a")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Micha\AppData\Local\Programs\Python\Python312\Lib\re\__init__.py", line 186, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Micha\AppData\Local\Programs\Python\Python312\Lib\re\__init__.py", line 334, in _compile_template
    return _sre.template(pattern, _parser.parse_template(repl, pattern))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Micha\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 1075, in parse_template
    raise s.error('bad escape %s' % this, len(this)) from None
re.error: bad escape \z at position 0
>>> "a".replace(r"\z")
>>> "a".replace("a", r"\z")
'\\z'
```

Reply:
And numbered groups:

```
re.sub(r"(a)", r"\1\1\1", "a")
'aaa'
```

Reply:
Reading code is hard. That's all covered because it uses an allowlist. Sorry